isocroft-syntax.md

```diff
@@ -0,0 +1,913 @@
+    
+## ARITHMETIC OPERATION
+- Operations in ratio can either be arithmetic, bitwise, logical, relational or functional
+
+####  functional opearator (binary)
+
+* `~>` (pipe)
+
+#### aritmetic operators (binary)
+
+* `+` (add)
+* `-` (subtract)
+* `*` (multiply)
+* `/` (divide)
+* `%` (modulo)
+* `+=` (rewrite add)
+* `-=` (rewrite subtract)
+
+#### arithmetic operators (unary)
+
+* `++` (increment)
+* `--` (decrement)
+
+#### logical operators (binary)
+
+* `&&` (and)
+* `and` (and)
+* `||` (or)
+* `or` (or)
+
+#### logical operators (unary)
+
+* `!` (not)
+* `not` (not)
+
+#### relational operators (binary)
+
+* `<`
+* `>`
+* `>=`
+* `<=`
+* `!=`
+* `!==`
+
+#### bitwise operators (binary)
+
+* `>>`
+
+```js
+
+var item = 3
+
+print(--item) // 2
+print(++item) // 3
+
+```
+
+## ARRAYS ( Mutable / Immutable )
+- Arrays in ratio will serve as an array of variables of unmixed or mixed types
+
+>mutable arrays (lists) are defined using `var`
+```js
+
+var int<4> list = [1, 2, 3, 4]
+
+list[0] = 0 // No error
+
+print(list); // 0, 2, 3, 4
+
+```
+
+>immutable arrays (tuples) are defined using `const`
+
+```js
+
+const int<2> tuple = [1, 5]; # you can't add or remove from the tuple but you can change the content at all indexes
+
+tuple[0] = 2 // No error
+
+print(tuple); // 2, 5
+```
+
+- Array of a complex type in ratio
+
+```js
+
+var Person<3> persons = [ 
+   {name: "Tochukwu Nkedilim", age: 27}, 
+   {name: "Ikechi Micheal", age: 29}, 
+   {name: "Azeez Adewale", age: 26} 
+]
+```
+
+## SEMI COLONS
+- Semi colons are used to separate multiple statements or expressions.
+
+    ```swift
+    const name = "John"; const age = 40;
+    print(name, age);
+    ```
+
+- Semi colons are optional if statements/expressions are separated by newlines.
+
+    ```cpp
+    const name = "John"
+    const age = 40
+    print(name, age)
+    ```
+
+## TYPE ANNOTATION
+- Types can be automatically inferred in Ratio.
+    ```js
+    var todo = "Create a new programming language :)" // `todo` is inferred as `str` (String)
+    ```
+
+    ```cpp
+    func add(a, b) {
+        return a + b
+    }
+
+    const sum = add(1, 1) // Usage of `add` inferred as `(int, int) -> int` here.
+    ```
+
+- Subjects can be explicitly type-annotated.
+
+    ```js
+    var int identification = 60
+    ```
+
+- Functions can also be explicitly type-annotated.
+
+    ```swift
+    func add(a int, b int) -> int {
+        return a + b
+    }
+    ```
+
+## CONCATENATION ( OVERLOADED OPERATOR - `++` )
+
+- The concatenation symbol for ratio is `++`
+
+```swift
+func sort(arr) -> int {
+   
+}
+```
+
+>the pipe operator `~>` can also be used to sort the resulting list
+```elixir
+
+var int<8> numbers = [1,3,5,7] ++ [0,2,4,6] ~> sort
+
+```
+
+- The concatenation symbol can also be used to join for a list of characters (strings) in ratio
+
+```elixir
+
+print("Hello " ++ "World");
+
+```
+
+## DESTRUCTURING
+- Destructuring a string.
+
+```js
+
+  const [...chars] = "Happy Birthday!"
```

Comment:
This is another example of the feature i spoke of. The literal ( *"Happy Birthday!"* ) on the RHS will be "reduced statically" into the LHS. on the AST, the "Happy Birthday!" will be "pre-executed" and we will have the below:

chars -> identifier
char<15> -> type
[ "H", "a", "p", "p", "y", " ", "B", "i", "r", "t", "h", "d", "a", "y", "!" ] -> value

Reply:
This feature allows users/programmers to write less verbose code. I am not naive and i believe this feature might be very difficult or challenging to implement yet however worth discussing and considering.

Reply:
Nice idea, however, it is going to be inconsistent with other destructuring syntax.

`...` is like a rest or spread syntax similar to what we have in JavaScript.

```swift
let [first, ...remaining] = [1, 2, 3]

first == 1 // type Int
remaining == [2, 3] // type List
```

So if at all we want that string destructuring syntax, then it should follow the same semantics.

```swift
let [first, ...remaining] = "abc"

first == 'a' // type Char
remaining == "bc" // type String
```

However, I have an issue with that syntax, because it is already reserved for list destructuring. It would be nice to find another _nice_ syntax for string destructuring.

Reply:
I agree @appcypher.

Ok i will work on that and bring my alternate proposal for *string destructuring* back here by tomorrow (Saturday [ 16th, Nov 2019 ]) evening so we can have further discussions.